stdlib/builtins.pyi

```diff
@@ -109,19 +110,19 @@ class object:
     def __dir__(self) -> Iterable[str]: ...
     def __init_subclass__(cls) -> None: ...
 
-class staticmethod(object):  # Special, only valid as a decorator.
-    __func__: Callable[..., Any]
+class staticmethod(Generic[_FuncT]):  # Special, only valid as a decorator.
+    __func__: _FuncT
     __isabstractmethod__: bool
-    def __init__(self, f: Callable[..., Any]) -> None: ...
+    def __init__(self, f: _FuncT) -> None: ...
     def __new__(cls: Type[_T], *args: Any, **kwargs: Any) -> _T: ...
-    def __get__(self, obj: _T, type: Optional[Type[_T]] = ...) -> Callable[..., Any]: ...
+    def __get__(self, obj: _T, type: Type[_T] | None = ...) -> _FuncT: ...
 
-class classmethod(object):  # Special, only valid as a decorator.
-    __func__: Callable[..., Any]
+class classmethod(Generic[_FuncT]):  # Special, only valid as a decorator.
+    __func__: _FuncT
     __isabstractmethod__: bool
-    def __init__(self, f: Callable[..., Any]) -> None: ...
+    def __init__(self, f: _FuncT) -> None: ...
     def __new__(cls: Type[_T], *args: Any, **kwargs: Any) -> _T: ...
-    def __get__(self, obj: _T, type: Optional[Type[_T]] = ...) -> Callable[..., Any]: ...
+    def __get__(self, obj: _T, type: Type[_T] | None = ...) -> _FuncT: ...
```

Comment:
This doesn't seem right, because the callable returned from `__get__` doesn't take the class as a first argument.

```python
>>> class Foo:
...     @classmethod
...     def bar(cls):
...             print("Bar runs:", cls)
... 
>>> Foo.bar()
Bar runs: <class '__main__.Foo'>
>>> Foo.__dict__['bar'].__func__(Foo)
Bar runs: <class '__main__.Foo'>
```

Here `Foo.__dict__['bar']` is the `classmethod` object, and for `Foo.bar`, its `__get__` method gets called. The result is a bound method, which basically means that it includes the class similarly to `partial`:

```python
>>> Foo.bar == Foo.__dict__['bar'].__get__(Foo(), None)
True
>>> Foo.bar
<bound method Foo.bar of <class '__main__.Foo'>>
```

Reply:
Another problem: I think it calls `__get__(None, Foo)` when looking up `Foo.bar`, so `None` should be allowed for the first argument to `__get__`.

Reply:
Good catch. I tried to express this using `ParamSpec`, but I'm not sure this is correct. Also, do you know in what circumstances the `owner` argument can be `None` or not set?

Reply:
It appears to be a technicality that doesn't really occur in practice: "Python’s own `__getattribute__()` implementation always passes in both arguments whether they are required or not." https://docs.python.org/3/reference/datamodel.html#object.__get__